### PR TITLE
Feature etb#95 adding hwas module

### PR DESCRIPTION
**Pull request by jrachucki-n7s**

### CODE REVIEW

File: src/hwas/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,16 @@
+project(Samv71RuntimeHwas VERSION 1.0.0 LANGUAGES C)
+
+add_library(Samv71RuntimeHwas STATIC)
+target_sources(Samv71RuntimeHwas
+    PRIVATE     hwas.c
+    PUBLIC      hwas.h)
+target_include_directories(Samv71RuntimeHwas
+    PRIVATE     ${CMAKE_SOURCE_DIR}/src
```
Comment (Glamhoth) [inline src/hwas/CMakeLists.txt]: `CMAKE_CURRENT_SOURCE_DIR`

File: src/hwas/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    PUBLIC      ${CMAKE_CURRENT_SOURCE_DIR}/dataview)
+target_link_libraries(Samv71RuntimeHwas
+    PRIVATE     SAMV71::Nvic)
+
+add_format_target(Samv71RuntimeHwas)                
+
+set_target_properties(Samv71RuntimeHwas PROPERTIES OUTPUT_NAME "hwas")
+add_library(SAMV71::Runtime::hwas ALIAS Samv71RuntimeHwas)
```
Comment (Glamhoth) [inline src/hwas/CMakeLists.txt]: `::Hwas`

File: src/hwas/hwas.h [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    void hwas_PI_InterruptSubscriptionManagement_SubscribeToInterrupt_Pi(const asn1SccInterruptNumber*);
+    void hwas_PI_RawMemoryAccess_ReadBuffer_Pi(const asn1SccSourceAddress*, asn1SccByteBuffer*);
+    void hwas_PI_RawMemoryAccess_ReadWord_Pi(const asn1SccSourceAddress*, const asn1SccWordMask*, asn1SccWord*);
+    void hwas_PI_RawMemoryAccess_WriteBuffer_Pi(const asn1SccDestinationAddress*, const asn1SccByteBuffer*);
+    void hwas_PI_RawMemoryAccess_WriteWord_Pi(const asn1SccDestinationAddress*,
+                                              const asn1SccWordMask*,
+                                              const asn1SccWord*);
+
+    /* Required interfaces */
+    extern void hwas_RI_InterruptSubscription_Interrupt_Ri(const asn1SccInterrupt_Type*);
```
Comment (Glamhoth) [inline src/hwas/hwas.h:41]: Formatting?

**Review by Glamhoth — CHANGES_REQUESTED** [review]

File: src/hwas/CMakeLists.txt (hunk 1 above)
Comment (jrachucki-n7s) [reply src/hwas/CMakeLists.txt]: fixed in new commit

File: src/hwas/CMakeLists.txt (hunk 2 above)
Comment (jrachucki-n7s) [reply src/hwas/CMakeLists.txt]: fixed in new commit

File: src/hwas/hwas.h (hunk 3 above)
Comment (jrachucki-n7s) [reply src/hwas/hwas.h:41]: Indentation forced by clang.

Comment (Glamhoth) [reply src/hwas/hwas.h:41]: ack

**Review by Glamhoth — DISMISSED** [review]

File: src/hwas/hwas.c [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+        asn1SccInterrupt_Type irq = { .interrupt = Nvic_Irq_Spi1 };
+        hwas_RI_InterruptSubscription_Interrupt_Ri(&irq);
+    }
+}
+
+void
+TWIHS0_Handler(void)
+{
+    if(interruptSubscribe[Nvic_Irq_Twihs0]) {
+        asn1SccInterrupt_Type irq = { .interrupt = Nvic_Irq_Twihs2 };
```
Comment (Lurkerpas) [inline src/hwas/hwas.c]: Twihs 0 vs 2

File: src/hwas/hwas.c [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+        hwas_RI_InterruptSubscription_Interrupt_Ri(&irq);
+    }
+}
+
+void
+MCAN0_Handler(void)
+{
+    if(interruptSubscribe[Nvic_Irq_Mcan0_Irq0]) {
+        asn1SccInterrupt_Type irq = { .interrupt = Nvic_Irq_Mcan0_Irq1 };
```
Comment (Lurkerpas) [inline src/hwas/hwas.c:135]: 0 vs 1

File: src/hwas/hwas.c [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+        hwas_RI_InterruptSubscription_Interrupt_Ri(&irq);
+    }
+}
+
+void
+MCAN1_Handler(void)
+{
+    if(interruptSubscribe[Nvic_Irq_Mcan0_Irq1]) {
+        asn1SccInterrupt_Type irq = { .interrupt = Nvic_Irq_Mcan1_Irq0 };
```
Comment (Lurkerpas) [inline src/hwas/hwas.c:144]: 0 vs 1

File: src/hwas/hwas.c [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+        address++;
+    }
+}
+
+void
+hwas_PI_RawMemoryAccess_ReadWord_Pi(const asn1SccSourceAddress* IN_address,
+                                    const asn1SccWordMask* IN_mask,
+                                    asn1SccWord* OUT_value)
+{
+    uint32_t addressInValue = (uint32_t)*IN_address;
```
Comment (Lurkerpas) [inline src/hwas/hwas.c]: Shouldn't you map *In_address to uint32_t*, which should be then read?

File: src/hwas/hwas.c [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+void
+hwas_PI_InterruptSubscriptionManagement_SubscribeToInterrupt_Pi(const asn1SccInterruptNumber* IN_interrupt)
+{
+    interruptSubscribe[*IN_interrupt] = true;
+}
+
+void
+hwas_PI_RawMemoryAccess_ReadBuffer_Pi(const asn1SccSourceAddress* IN_address, asn1SccByteBuffer* OUT_buffer)
+{
+    uint32_t* address = (uint32_t*)IN_address;
```
Comment (Lurkerpas) [inline src/hwas/hwas.c]: *

File: src/hwas/hwas.c [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+    uint32_t addressInValue = (uint32_t)*IN_address;
+    uint32_t maskValue = (uint32_t)*IN_mask;
+    uint32_t* addressOut = (uint32_t*)OUT_value;
+    *addressOut = maskValue & addressInValue;
+}
+
+void
+hwas_PI_RawMemoryAccess_WriteBuffer_Pi(const asn1SccDestinationAddress* IN_address, const asn1SccByteBuffer* IN_buffer)
+{
+    uint32_t* address = (uint32_t*)IN_address;
```
Comment (Lurkerpas) [inline src/hwas/hwas.c]: *

File: src/hwas/hwas.c (hunk 4 above)
Comment (jrachucki-n7s) [reply src/hwas/hwas.c]: fixed in new commit

File: src/hwas/hwas.c (hunk 5 above)
Comment (jrachucki-n7s) [reply src/hwas/hwas.c:135]: fixed in new commit

File: src/hwas/hwas.c (hunk 6 above)
Comment (jrachucki-n7s) [reply src/hwas/hwas.c:144]: fixed in new commit

File: src/hwas/hwas.c (hunk 7 above)
Comment (jrachucki-n7s) [reply src/hwas/hwas.c]: fixed in new commit

File: src/hwas/hwas.c (hunk 8 above)
Comment (jrachucki-n7s) [reply src/hwas/hwas.c]: fixed in new commit

File: src/hwas/hwas.c (hunk 9 above)
Comment (jrachucki-n7s) [reply src/hwas/hwas.c]: fixed in new commit

File: src/hwas/hwas.c [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+void
+hwas_startup(void)
+{
+    memset((void*)interruptSubscribe, 0, Nvic_InterruptCount);
+}
+
+void
+hwas_PI_InterruptManagement_DisableInterrupt_Pi(const asn1SccInterruptNumber* IN_interrupt)
+{
+    Nvic_disableInterrupt((Nvic_Irq)IN_interrupt);
```
Comment (Lurkerpas) [inline src/hwas/hwas.c]: *In_Interrupt

File: src/hwas/hwas.c [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+}
+
+void
+hwas_PI_InterruptManagement_EnableInterrupt_Pi(const asn1SccInterruptNumber* IN_interrupt)
```
Comment (Lurkerpas) [inline src/hwas/hwas.c:166]: *In_Interrupt

**Review by Lurkerpas — DISMISSED** [review]

**Review by Lurkerpas — APPROVED** [review]

**Pull request merged**